### ISSUE TITLE
compression: GET and HEAD headers do not match

### ISSUE THREAD

**Issue by SaveTheRbtz** [envoyproxy/envoy#14225]
*Title*: Content-Encoding header is not added to HEAD responses.

*Description*:
Following header is not present in HEAD responses:
```content-encoding: gzip```

[RFC7231](https://tools.ietf.org/html/rfc7231#section-4.3.2) specifies the following:
> The server SHOULD send the same
>    header fields in response to a HEAD request as it would have sent if
>    the request had been a GET, except that the payload header fields
>    (Section 3.3) MAY be omitted.

(NB! Content-Encoding is not in the list of payload-specific headers.)

Therefore this may cause issues with clients that follow RFC and expect HEAD and GET headers to match.

cc: @rojkov @antoniovicente

**Comment by antoniovicente** [comment]
Could you provide some details about how the envoy is configured?  Is the response compression done by the Envoy by means of a filter or by the upstream that generated the response?

**Comment by SaveTheRbtz** [comment]
Yes, compression is done by Envoy.  The setup is pretty much the same as our nginx:
```python
def default_compressor_config(
    compression_level: Gzip.CompressionLevel = Gzip.CompressionLevel.DEFAULT_COMPRESSION,
) -> Compressor:
    return Compressor(
        # Envoy's default is 30.  Nginx uses 20.
        content_length=UInt32Value(value=128),
        remove_accept_encoding_header=False,
        content_type=default_compressible_mime_types(),
        compressor_library=typed_extension_config(
            name="default",
            config=Gzip(
                # Envoy's default is 6 (Z_DEFAULT_COMPRESSION).  Nginx uses a mix of 6 and 1.
                compression_level=compression_level,
                # Envoy's default is 4k (12 bits).  Nginx uses 32k (MAX_WBITS, 15 bits).
                window_bits=UInt32Value(value=12),
                # Envoy's default is 5.  Nginx uses 8 (MAX_MEM_LEVEL - 1).
                memory_level=UInt32Value(value=5),
            ),
        ),
    )
```

This behaviour change can be observed live (same backend):

Nginx:
```
$ curl -I --compressed -s --resolve www.dropbox.com:443:162.125.248.1 https://www.dropbox.com/404
```
Envoy:
```
$ curl -I --compressed -s --resolve www.dropbox.com:443:162.125.248.18 https://www.dropbox.com/404
```

**Comment by antoniovicente** [comment]
ok, that makes sense.  I wouldn't be surprised if there were other more fundamental differences like the response to HEAD having a content-length and the GET having no content-length(and in the case of HTTP/1.1 transfer-encoding: chuncked) or a different content length.

Minimizing these differences is relatively challenging.

**Comment by rojkov** [comment]
I guess the problem is that in `CompressorFilter::encodeHeaders()` we just skip setting `content-encoding`  for all body-less responses. Probably there are two options here:

1. modify headers even if there's no body;
2. don't touch `content-encoding` at all, but update `transfer-encoding` instead.

The latter seems to adhere to the spec better (https://issues.apache.org/bugzilla/show_bug.cgi?id=39727#c31), but not all clients comply with this part of the spec.

**Comment by toddmgreer** [comment]
The compression filter should use the same logic on HEAD and GET responses headers to decide whether a response should be compressed, and should then modify the response headers in the same way, including setting `content-encoding`. The only difference should be lack of an actual body.

It may be that the response headers from upstream differ between HEAD and GET responses in a way that produces different outcomes, but there's nothing the compression filter can do about that.

Re updating `transfer-encoding`: because `transfer-encoding` is malformed in H/2[1], and Envoy is H/2-focused, I suspect that would be problematic.

[1] https://httpwg.org/specs/rfc9113.html#rfc.section.8.2.2.
